# Jupiter Notebook for Groupwork in AA 20224

#### Import Libaries

In [167]:
import os
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from IPython.display import display
from datetime import timedelta

### 1 Load and Explore Data:

In [168]:
# Load csv-files independently of the operating system (mac or windows or linux) in a pandas dataframe
charging_data = pd.read_csv(os.path.join('Data', 'charging_sessions.csv'))
weather_data = pd.read_csv(os.path.join('Data', 'weather_burbank_airport.csv'))

try:
    print("Info Data of charging_data Dataframe:")
    display(charging_data.info())
except Exception as e:
    print("Error in charging_data.info():", e)


Info Data of charging_data Dataframe:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        66450 non-null  int64  
 1   id                66450 non-null  object 
 2   connectionTime    66450 non-null  object 
 3   disconnectTime    66450 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      66450 non-null  float64
 6   sessionID         66450 non-null  object 
 7   siteID            66450 non-null  int64  
 8   spaceID           66450 non-null  object 
 9   stationID         66450 non-null  object 
 10  timezone          66450 non-null  object 
 11  userID            49187 non-null  float64
 12  userInputs        49187 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.6+ MB


None

In [169]:
try:
    print("Info Data of weather_data Dataframe")
    display(weather_data.info())
except Exception as e:
    print("Error in weather_data.info():", e)

Info Data of weather_data Dataframe
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29244 entries, 0 to 29243
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   city                     29244 non-null  object 
 1   timestamp                29244 non-null  object 
 2   temperature              29219 non-null  float64
 3   cloud_cover              29224 non-null  float64
 4   cloud_cover_description  29224 non-null  object 
 5   pressure                 29236 non-null  float64
 6   windspeed                29158 non-null  float64
 7   precipitation            29244 non-null  float64
 8   felt_temperature         29218 non-null  float64
dtypes: float64(6), object(3)
memory usage: 2.0+ MB


None

In [170]:
try:
    print("charging_data first few rows:")
    display(charging_data.head())
except Exception as e:
    print("Error in charging_data.head():", e)

charging_data first few rows:


Unnamed: 0                        id             connectionTime  \
0           0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1           1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2           2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3           3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4           4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   

              disconnectTime           doneChargingTime  kWhDelivered  \
0  2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1  2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2  2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3  2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4  2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                          userInputs  
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...

In [171]:
try:
    print("weather_data first few rows:")
    display(weather_data.head())
except Exception as e:
    print("Error in weather_data.head():", e)

weather_data first few rows:


city            timestamp  temperature  cloud_cover  \
0  Burbank  2018-01-01 08:53:00          9.0         33.0   
1  Burbank  2018-01-01 09:53:00          9.0         33.0   
2  Burbank  2018-01-01 10:53:00          9.0         21.0   
3  Burbank  2018-01-01 11:53:00          9.0         29.0   
4  Burbank  2018-01-01 12:53:00          8.0         33.0   

  cloud_cover_description  pressure  windspeed  precipitation  \
0                    Fair    991.75        9.0            0.0   
1                    Fair    992.08        0.0            0.0   
2                    Haze    992.08        0.0            0.0   
3           Partly Cloudy    992.08        0.0            0.0   
4                    Fair    992.08        0.0            0.0   

   felt_temperature  
0               8.0  
1               9.0  
2               9.0  
3               9.0  
4               8.0

In [172]:
try:
    print("charging_data description:")
    display(charging_data.describe())
except Exception as e:
    print("Error in charging_data.describe():", e)

charging_data description:


Unnamed: 0  kWhDelivered        siteID        userID
count  66450.000000  66450.000000  66450.000000  49187.000000
mean    4726.102062     11.787916      1.472656   1800.494582
std     3568.025394     10.314789      0.499256   2617.548614
min        0.000000      0.501000      1.000000      1.000000
25%     1845.000000      5.097000      1.000000    431.000000
50%     3960.000000      9.140000      1.000000    818.000000
75%     7027.000000     14.183000      2.000000   1805.000000
max    15291.000000    108.797242      2.000000  19923.000000

In [173]:
try:
    print("weather_data description:")
    display(weather_data.describe())
except Exception as e:
    print("Error in weather_data.describe():", e)

weather_data description:


temperature   cloud_cover      pressure     windspeed  precipitation  \
count  29219.000000  29224.000000  29236.000000  29158.000000   29244.000000   
mean      17.885622     30.107993    986.879400      8.566568       0.059360   
std        6.454604      5.400388      3.607065      6.699889       0.471682   
min        2.000000      4.000000    971.000000      0.000000       0.000000   
25%       13.000000     26.000000    984.500000      6.000000       0.000000   
50%       17.000000     33.000000    986.480000      7.000000       0.000000   
75%       22.000000     34.000000    989.110000     13.000000       0.000000   
max       46.000000     47.000000    999.650000     57.000000      18.540000   

       felt_temperature  
count      29218.000000  
mean          17.731433  
std            6.321100  
min            0.000000  
25%           13.000000  
50%           17.000000  
75%           22.000000  
max           42.000000

#### Transform Datatyp

In [174]:
#format columns into datetime-objects
charging_data['connectionTime'] = pd.to_datetime(charging_data['connectionTime'])
charging_data['disconnectTime'] = pd.to_datetime(charging_data['disconnectTime'])
charging_data['doneChargingTime'] = pd.to_datetime(charging_data['doneChargingTime'])
#Todo also for the rest (inf needed) -> Chiara
#check for type of userID
userID_type = charging_data['userID'].dtype
print("data type of userID: ", userID_type)
#float64

#check for number of wrong  type within column userID
userID_wrong_type = charging_data['userID'].apply(lambda x: not isinstance(x, float)).sum()
print("number of incorrect data types: ", userID_wrong_type)
#-->no wrong types


#Todo also for the weather data -> Chiara

display(charging_data.info())

data type of userID:  float64
number of incorrect data types:  0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype              
---  ------            --------------  -----              
 0   Unnamed: 0        66450 non-null  int64              
 1   id                66450 non-null  object             
 2   connectionTime    66450 non-null  datetime64[ns, UTC]
 3   disconnectTime    66450 non-null  datetime64[ns, UTC]
 4   doneChargingTime  62362 non-null  datetime64[ns, UTC]
 5   kWhDelivered      66450 non-null  float64            
 6   sessionID         66450 non-null  object             
 7   siteID            66450 non-null  int64              
 8   spaceID           66450 non-null  object             
 9   stationID         66450 non-null  object             
 10  timezone          66450 non-null  object             
 11  userID            49187 non-null  float64            


None

### 2 Check and Handle missing Data

#### 2.1 Identifying missing data

In [175]:
#Identifying and returning missing data:
charging_missing_data = charging_data.isnull().sum()
display(charging_missing_data)

weather_missing_data = weather_data.isnull().sum()
display(weather_missing_data[weather_missing_data > 0])
print("Number of unique lines with missing values: " + str(weather_data.isnull().any(axis=1).sum()))


Unnamed: 0              0
id                      0
connectionTime          0
disconnectTime          0
doneChargingTime     4088
kWhDelivered            0
sessionID               0
siteID                  0
spaceID                 0
stationID               0
timezone                0
userID              17263
userInputs          17263
dtype: int64

temperature                25
cloud_cover                20
cloud_cover_description    20
pressure                    8
windspeed                  86
felt_temperature           26
dtype: int64

Number of unique lines with missing values: 139


#### 2.2 Handling missing data

**Lets look at the missing values**<br>
For `charging_data`:
- `doneChargingTime`: We have 4088 missing values, that is 4088/66450=6,15%. Because we do not want to loose 6% of the data, we will replace the missing values with connectionTime + median charging time.
- `userID`: We have 17263 missing values, that is 17263/66450=25,98%. We will replace them with null values (0.0).
- `userInput`: We have 17263 missing values, that is 17263/66450=25,98%. We will replace them with null values (Null).

For `weather_data`:
- There are 139 lines with missing values through the data set. Because this is only 0,47%, we decided to drop these records. 

In [176]:
#doneChargingTime 
#baisc idea: replace missing values with connectionTime + median of chargingTime
#store values from doneChargingTime into new column doneChargingTime_clean (can be transferred to doneChargingTime) in order to load them into a new csv

#calculate median_charging_time in whole seconds of all non-null rows (parameter skip_na of median()-method is true by default)
median_charging_time = (
    (charging_data['doneChargingTime'] - charging_data['connectionTime']).dt.total_seconds()).median().round()

#calculate new doneChargingTime based on median_charging_time
calculated_doneChargingTime = charging_data['connectionTime'] + timedelta(seconds=median_charging_time)
#add new column where null values are filled with calculated_doneChargingTime
charging_data['doneChargingTime_clean'] = charging_data['doneChargingTime'].fillna(calculated_doneChargingTime)

#check doneChargingTime_clean (can be deleted later)-->
print(median_charging_time / 60 / 60)
# print(calculated_doneChargingTime.iloc[1])
# print(charging_data['connectionTime'].iloc[1])
# we need to check if the doneChargingTime_clean is now after the disconnectTime
tmp_check_doneChargingTime = charging_data[
    charging_data['doneChargingTime'].isnull() & (charging_data['doneChargingTime_clean'] > charging_data['disconnectTime'])]
print(tmp_check_doneChargingTime[['doneChargingTime', 'doneChargingTime_clean', 'disconnectTime']])
# TODO: entpsrechende Einträge korrigieren oder in Zeile 15 schon abändern... oder vielleicht erst unter erroneous Data behandeln, um die anderen Einträge noch mitzunehmen?


2.783888888888889
      doneChargingTime    doneChargingTime_clean            disconnectTime
3759               NaT 2020-04-09 16:48:55+00:00 2020-04-09 16:33:40+00:00
3763               NaT 2020-04-10 16:05:01+00:00 2020-04-10 13:49:37+00:00
3765               NaT 2020-04-10 17:42:09+00:00 2020-04-10 15:43:03+00:00
3766               NaT 2020-04-10 19:08:17+00:00 2020-04-10 18:16:22+00:00
3767               NaT 2020-04-10 21:41:00+00:00 2020-04-10 20:37:47+00:00
...                ...                       ...                       ...
58435              NaT 2019-02-17 03:26:18+00:00 2019-02-17 03:08:59+00:00
60411              NaT 2019-03-30 03:19:12+00:00 2019-03-30 02:23:14+00:00
61103              NaT 2019-04-15 15:48:01+00:00 2019-04-15 13:55:25+00:00
61117              NaT 2019-04-15 16:39:47+00:00 2019-04-15 14:47:01+00:00
65530              NaT 2019-07-13 23:00:20+00:00 2019-07-13 20:37:40+00:00

[1744 rows x 3 columns]


In [177]:
#userID
# TODO: müsste die userID nicht eigentlich ein string sein und kein float? (unabh. davon, dass wir die Spalte hinterher ja eh converten wollten)
#replace missing values with 0.0 (float)
#store values from userID into new column userID_clean (can be transferred to userID)
charging_data['userID_clean'] = charging_data['userID'].fillna(0.0)

#check userID_clean for consistency (can be deleted later) -->
userID_wrong_typ = charging_data['userID_clean'].apply(lambda x: not isinstance(x, float)).sum()
userID_wrong_typ

comparison = charging_data[charging_data['userID'].isnull()][['userID_clean', 'userID' ]]
comparison['userID_clean_type'] = charging_data['userID_clean'].dtype
print(comparison)


       userID_clean  userID userID_clean_type
11              0.0     NaN           float64
41              0.0     NaN           float64
180             0.0     NaN           float64
193             0.0     NaN           float64
265             0.0     NaN           float64
...             ...     ...               ...
66271           0.0     NaN           float64
66316           0.0     NaN           float64
66321           0.0     NaN           float64
66415           0.0     NaN           float64
66449           0.0     NaN           float64

[17263 rows x 3 columns]


In [178]:
# drop rows with missing information of weather_data
weather_data = weather_data.dropna(ignore_index=True)

# check if it worked/ how many rows remained
display(weather_data.info())



# ToDo -> New df for cleaned data -> Paul

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29105 entries, 0 to 29104
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   city                     29105 non-null  object 
 1   timestamp                29105 non-null  object 
 2   temperature              29105 non-null  float64
 3   cloud_cover              29105 non-null  float64
 4   cloud_cover_description  29105 non-null  object 
 5   pressure                 29105 non-null  float64
 6   windspeed                29105 non-null  float64
 7   precipitation            29105 non-null  float64
 8   felt_temperature         29105 non-null  float64
dtypes: float64(6), object(3)
memory usage: 2.0+ MB


None

### 3 Erroneous Data
### ToDo User Input Data
Format UserInputData into an array?

In [179]:
#ToDo -> Fynn

### ToDo Data correction: Disconnect time is before done charging time
The data shows that is some cases Disconnect time is before done charging time...thats an error...Solution If(DoneTime after DiscTime): DoneTime == DiscTime

In [180]:
# Check whether disconnectTime is before connectionTime
erroneous_disconnect_times = charging_data[charging_data['disconnectTime'] < charging_data['connectionTime']]

# Output the wrong lines
print(erroneous_disconnect_times[['connectionTime', 'disconnectTime']])


Empty DataFrame
Columns: [connectionTime, disconnectTime]
Index: []


In [181]:
# Calculate the loading time in seconds
charging_data['charging_duration'] = (charging_data['doneChargingTime_clean'] - charging_data['connectionTime']).dt.total_seconds()

# Calculate the total connection time in seconds
charging_data['connection_duration'] = (charging_data['disconnectTime'] - charging_data['connectionTime']).dt.total_seconds()

# Check whether the loading time is longer than the connection time
erroneous_charging_times = charging_data[charging_data['charging_duration'] > charging_data['connection_duration']]

# Count the number of rows with erroneous charging times
erroneous_count = erroneous_charging_times.shape[0]

# Display the count
print(f"Number of rows with longer charging duration then connection: {erroneous_count}")

print(erroneous_charging_times[['connectionTime', 'disconnectTime', 'doneChargingTime_clean', 'charging_duration', 'connection_duration']])

Number of rows with longer charging duration then connection: 6436
                 connectionTime            disconnectTime  \
12    2020-01-02 15:04:38+00:00 2020-01-02 22:08:39+00:00   
20    2020-01-02 15:28:47+00:00 2020-01-02 19:01:54+00:00   
25    2020-01-02 15:42:05+00:00 2020-01-02 21:58:45+00:00   
26    2020-01-02 15:57:24+00:00 2020-01-02 16:35:37+00:00   
33    2020-01-02 16:34:35+00:00 2020-01-02 18:49:41+00:00   
...                         ...                       ...   
66432 2019-07-31 14:36:02+00:00 2019-07-31 22:34:10+00:00   
66439 2019-07-31 14:45:02+00:00 2019-07-31 23:43:20+00:00   
66440 2019-07-31 14:45:29+00:00 2019-07-31 22:01:32+00:00   
66445 2019-07-31 18:08:04+00:00 2019-07-31 23:29:18+00:00   
66447 2019-07-31 19:04:40+00:00 2019-07-31 22:44:22+00:00   

         doneChargingTime_clean  charging_duration  connection_duration  
12    2020-01-02 22:09:36+00:00            25498.0              25441.0  
20    2020-01-02 19:02:51+00:00            12844.0  

In [182]:
#Identify the differance between the 6436 changed data and the 6441 equal datas one underneath
# Identify lines where doneChargingTime_clean was already equal to disconnectTime
pre_existing_equal_times = charging_data[
    charging_data['doneChargingTime_clean'] == charging_data['disconnectTime']
]

# Count how many of them were marked as faulty
pre_existing_errors = pre_existing_equal_times[
    pre_existing_equal_times['charging_duration'] > pre_existing_equal_times['connection_duration']
]

print(f"Number of rows where doneChargingTime_clean was already equal to disconnectTime: {pre_existing_equal_times.shape[0]}")
print(f"Number of these rows identified as erroneous: {pre_existing_errors.shape[0]}")

Number of rows where doneChargingTime_clean was already equal to disconnectTime: 5
Number of these rows identified as erroneous: 0


In [183]:
#Adjust doneChargingTime_clean if it exceeds disconnectTime
# Set doneChargingTime_clean to disconnectTime where it exceeds disconnectTime
charging_data.loc[charging_data['doneChargingTime_clean'] > charging_data['disconnectTime'], 'doneChargingTime_clean'] = charging_data['disconnectTime']

# Verify adjustments
adjusted_charging_times = charging_data[charging_data['doneChargingTime_clean'] == charging_data['disconnectTime']]
print(f"Number of rows where doneChargingTime_clean was adjusted to disconnectTime: {adjusted_charging_times.shape[0]}")

print(adjusted_charging_times[['connectionTime', 'disconnectTime', 'doneChargingTime_clean']].head())

Number of rows where doneChargingTime_clean was adjusted to disconnectTime: 6441
              connectionTime            disconnectTime  \
12 2020-01-02 15:04:38+00:00 2020-01-02 22:08:39+00:00   
20 2020-01-02 15:28:47+00:00 2020-01-02 19:01:54+00:00   
25 2020-01-02 15:42:05+00:00 2020-01-02 21:58:45+00:00   
26 2020-01-02 15:57:24+00:00 2020-01-02 16:35:37+00:00   
33 2020-01-02 16:34:35+00:00 2020-01-02 18:49:41+00:00   

      doneChargingTime_clean  
12 2020-01-02 22:08:39+00:00  
20 2020-01-02 19:01:54+00:00  
25 2020-01-02 21:58:45+00:00  
26 2020-01-02 16:35:37+00:00  
33 2020-01-02 18:49:41+00:00  


### 4 Outliers

In [184]:
# Unrealistic Time Differences
# Check if the charging duration is either too short (less than 1 minute) or too long (more than 1 day).

unrealistic_times = charging_data[(charging_data['charging_duration'] < 60) | (charging_data['charging_duration'] > 86400)]

# Show the number of rows with unrealistic charging durations
print(f"Number of rows with unrealistic charging times: {unrealistic_times.shape[0]}")

# Show examples of the data
print("Examples of erroneous charging durations (too short (1 Min) or too long(24 hours)):")
print(unrealistic_times[['connectionTime', 'disconnectTime', 'doneChargingTime_clean', 'charging_duration']].head())


# ToDo: How to handle outliers? How to set the boundries? Is it necassary to handle the outliers (bzw. how much)? -Justin

Number of rows with unrealistic charging times: 95
Examples of erroneous charging durations (too short (1 Min) or too long(24 hours)):
                 connectionTime            disconnectTime  \
18604 2021-07-16 15:45:30+00:00 2021-07-19 23:06:05+00:00   
18606 2021-07-16 16:23:01+00:00 2021-07-19 23:05:42+00:00   
18607 2021-07-16 16:40:11+00:00 2021-07-19 23:05:30+00:00   
18609 2021-07-16 17:52:18+00:00 2021-07-19 20:41:13+00:00   
20425 2019-01-10 01:51:20+00:00 2019-01-11 07:17:28+00:00   

         doneChargingTime_clean  charging_duration  
18604 2021-07-19 20:36:25+00:00           276655.0  
18606 2021-07-19 20:45:36+00:00           274955.0  
18607 2021-07-19 20:45:36+00:00           273925.0  
18609 2021-07-19 20:41:07+00:00           269329.0  
20425 2019-01-11 07:17:28+00:00           106027.0  



### 5 Inconsistancies
Correct missspelling and false dataTypes

In [185]:

#data['session_date'] = pd.to_datetime(data['session_date'])  # Convert to datetime

#ToDo: Check missspelling etc., dataTypes have been done above -> Paul

### 6 Final review
Look again for missing, errorous, and inconsistant data

In [186]:
# Todo: Code